### PR TITLE
Generate IDA structures to resolve vtable and structure offsets

### PR DESCRIPTION
**Pull request by nneonneo**
This patch augments the IDA output module to add structure, function and TypeInfo types, which enable IDA to automatically resolve structure offsets and improve the decompilation massively. Take a look at the following decompiler sample (IDA 7.0, x86) for a pretty much randomly chosen function:

**Before**:
```c
// Int32 GetParametersCount()
int __cdecl MethodBase_GetParametersCount(int a1)
{
  int v1; // eax

  v1 = (*(int (__cdecl **)(int, _DWORD))(*(_DWORD *)a1 + 484))(a1, *(_DWORD *)(*(_DWORD *)a1 + 488));
  if ( !v1 )
    sub_100CCC90();
  return *(_DWORD *)(v1 + 12);
}
```

**After** (`v1` manually annotated with `ParameterInfo__Array *` type)
```c
// Int32 GetParametersCount()
int32_t __cdecl MethodBase_GetParametersCount(struct MethodBase *this)
{
  ParameterInfo__Array *v1; // eax

  v1 = (ParameterInfo__Array *)((int (__cdecl *)(struct MethodBase *, const void *))this->klass->vtable.GetParametersInternal.methodPtr)(
                                 this,
                                 this->klass->vtable.GetParametersInternal.method);
  if ( !v1 )
    sub_100CCC90();
  return v1->max_length;
}
```

Across the board, this patch enables IDA to automatically pick up static fields (via typed `TypeInfo` structs), regular fields (via typed value and reference class structures), vtables (via `TypeInfo` structs), and even interface calls if the relevant `InterfaceInvoker` code has been inlined. By typing all functions, it produces significantly better code for direct function calls, and improves type inference.

Unfortunately, I'm not that proficient with the internals of Il2CppInspector or with Il2Cpp metadata, so there are a few current limitations to be aware of:

- vtables have no types. This is probably an easy fix, except for:
- generics basically don't work at all. I could not figure out how to *easily* monomorphize generics, especially genericized vtables. Most of this is just me not fully understanding how generics work. This unfortunately breaks generic fields and makes their offsets wrong, so it's definitely something that needs to be fixed.
- this requires the `Il2CppClass` structure for the specific version of Il2Cpp being targeted. I don't have access to the headers for historic versions of Il2Cpp, so I couldn't add their headers.
- this is untested on anything except MSVC-produced x86/x64 .dlls on Windows. More testing is needed for different environments (particularly, calling conventions might need adjusting on other platforms).

If anyone could provide a little nudge towards any of these issues, I'll happily work on them. I'm leaving this as a draft right now but would appreciate feedback at this stage on anything (code quality, general approach, willingness to merge, etc.) so I can make adjustments.

### CODE REVIEW

**Comment by nneonneo** [comment]
(pinging @carterbush, @StijnSimons since they seemed interested in IDA output)

**Comment by djkaty** [comment]
Wow that's awesome, adding struct definitions had been on my mind but I've been wading through the IDAPython "documentation" blind and it's pretty awful. Thanks very much for taking the time to implement this.

I'll have to look at the actual code in more detail tomorrow. I have the majority of the il2cpp versions with the Il2CppClass definitions, any idea how we can implement that neatly? I also have a pile of ARM and A64 binaries for Android and iOS I can send over if you're willing to try and figure out the calling conventions etc.

Generics in IL are (at least to my little brain) really complicated and it took me a long time to figure out how to analyse and model them correctly(ish). A lot of Il2Cpp's structures are lightly modified versions of actual IL metadata. This book was extremely useful, particular with regards to generics: https://www.amazon.com/Expert-NET-Assembler-Serge-Lidin/dp/1590596463/ref=sr_1_1?crid=15AQYKBIY9JRS&keywords=expert+.net+2.0+il+assembler&qid=1582585543&s=books&sprefix=expert+.net+il%2Cstripbooks-intl-ship%2C222&sr=1-1

On that note though, you can't set the Definition property on a monomorphized generic type (or closed generic type to use the C# parlance). Il2Cpp's TypeDefinition is a direct equivalent of the IL TypeDef metadata. This is only used for non-generic types or generic type definitions (in C# this is specifically a generic type where none of the generic type parameters have been specialized).

Closed generic types that actually have code generated for them in the binary file come from IL TypeSpec metadata. In Il2Cpp, this is rolled into MethodSpec which is a related IL metadata item for closed generic methods, so in Il2Cpp, a MethodSpec can refer to a specialization of a generic type definition or a generic method definition.

All in all, this is just the tip of the iceberg on a very convoluted web of metadata tables and sometimes oblique jargon. If you want to access all of the closed generic types available in the analysed binary file, the easiest way to do this is basically to cheat and use the scummy Il2CppModel.TypesByMethodSpecClassIndex property then just iterate over them all. All of the TypeInfos in this Dictionary will have specialized type parameters.

I'll write some more tomorrow when I've had chance to look over your code properly and try it out!

**Comment by nneonneo** [comment]
> Wow that's awesome, adding struct definitions had been on my mind but I've been wading through the IDAPython "documentation" blind and it's pretty awful. Thanks very much for taking the time to implement this.

As you can see from the code, I actually cheated and just used `idc.parse_decls` a lot (plus `SetType` for functions). These two together make up a good chunk of my IDA workflow anyway, and let me not deal with the god-awful `struc` interface. (Plus, AFAICT they actually support more options and syntax than is reachable with `struc`, at least in IDA 7.0+).

> I'll have to look at the actual code in more detail tomorrow. I have the majority of the il2cpp versions with the Il2CppClass definitions, any idea how we can implement that neatly? I also have a pile of ARM and A64 binaries for Android and iOS I can send over if you're willing to try and figure out the calling conventions etc.

The older class definitions AFAICT used a `* vtable` pointer so there will need to be some slight adjustments there (not implemented in the current code because I didn't know which metadata version corresponded to the headers I was looking at). I wonder if maybe some resources keyed on the metadata version would be a nice way to do it? The headers I'm adding to IDA are just lightly modified from the real thing.

I would definitely love to look at those binaries. You can send me some kind of share link to [github username]@gmail.com. Having corresponding C# code would be helpful too (if they're OSS) for comparison.

> Generics in IL are (at least to my little brain) really complicated and it took me a long time to figure out how to analyse and model them correctly(ish). A lot of Il2Cpp's structures are lightly modified versions of actual IL metadata. This book was extremely useful, particular with regards to generics: https://www.amazon.com/Expert-NET-Assembler-Serge-Lidin/dp/1590596463/ref=sr_1_1?crid=15AQYKBIY9JRS&keywords=expert+.net+2.0+il+assembler&qid=1582585543&s=books&sprefix=expert+.net+il%2Cstripbooks-intl-ship%2C222&sr=1-1

Yeah, I had a lot to go through with the vtables already so I didn't spend much time on generics (relatively). I will probably take another whack at trying to understand them. The main thing that needs to be done is linking vtables of generic classes (and interfaces). 

> On that note though, you can't set the Definition property on a monomorphized generic type (or closed generic type to use the C# parlance). Il2Cpp's TypeDefinition is a direct equivalent of the IL TypeDef metadata. This is only used for non-generic types or generic type definitions (in C# this is specifically a generic type where none of the generic type parameters have been specialized).
> 
> Closed generic types that actually have code generated for them in the binary file come from IL TypeSpec metadata. In Il2Cpp, this is rolled into MethodSpec which is a related IL metadata item for closed generic methods, so in Il2Cpp, a MethodSpec can refer to a specialization of a generic type definition or a generic method definition.

This is exactly why I wanted to post my code in an early state - because I was certain I was doing at least one thing wrong :). The naming confused me, so I'm glad you cleared this up. I can probably take another whack at generics with this new context.

> All in all, this is just the tip of the iceberg on a very convoluted web of metadata tables and sometimes oblique jargon. If you want to access all of the closed generic types available in the analysed binary file, the easiest way to do this is basically to cheat and use the scummy Il2CppModel.TypesByMethodSpecClassIndex property then just iterate over them all. All of the TypeInfos in this Dictionary will have specialized type parameters.
> 
> I'll write some more tomorrow when I've had chance to look over your code properly and try it out!

Thanks! I'll also keep hacking on it, but the basic structure is all there now.

**Comment by djkaty** [comment]
I'm still working on issue 21 (unrelated so not tagging it) and there's only one of me hehe, I'll build you some Android stuff (the source files in Il2CppTests) and send it to you as soon as I get the time. I don't have the capability to build iOS binaries so for that you'll have to make do with a random assortment of free games that I use in my test suite, hope that's okay!

**Comment by djkaty** [comment]
Alright, check your email!

I've sent you two archives:

1. all of the il2cpp toolchains I have at hand. They move the class definitions between files quite a bit so I like to just open the entire folder in VS Code and do a find in files for whatever definition I need. You can get every version of Il2CppClass in here.

2. sample il2cpp applications. Those in folders named like GameAssembly-X-x86/64 are Windows standalone PE files; those just named X are Android ARMv7 (32-bit) builds. All of these come from the test source code in Il2CppTests\TestSources (i use the generate-binaries.ps1 script included in the repo to build them). The other folders - named with 'example' in them - are applications in the wild for various platforms.

To make your life a bit easier, you can extract the samples archive into a folder called TestBinaries which should be a sibling of TestSources and TestExpectedResults. At the PowerShell prompt, navigate to the Il2CppTests folder and run `generate-tests.ps1`. This will generate a test for each sample. Rebuild the solution in Visual Studio and you'll be able to find all of the samples as individual tests in Test Explorer. Then you can just right-click on the file of interest and Run/Debug as desired.

I've just pushed a commit which will make the test runner output the IDA script files as well as the C# stubs, to help with this :)

(the tests will fail btw but you don't need to worry about that. If the red mark irritates you and you want them to succeed, run `update-expected-results.ps1` after you've run all of the tests at least once and it will copy the current set of .cs files to TestExpectedResults, which is a folder of .cs files it compares with the actual results to confirm I haven't introduced any regressions when coding.

I've pulled your code but haven't looked at it yet. Hope this gives you something to chew on in the meantime 😎

**Comment by djkaty** [comment]
Hey,

Sorry for taking a bit to reply. I haven't had time to look over the new commits, I did spend a bit going over what was in the PR from before the weekend.

I'm going to say straight up that I do not 100% understand everything going on in this code, but I understand the end goal, so I'm somewhat trusting you know what you're doing and may have dumb questions :)

There are some small code style issues but I'm not going to nitpick over them (I'll fix them myself later) and overall the code style was very inline with the existing code, so really good job on that, I really appreciate the effort.

My first key question is about `populateConcreteImplementations`. An abstract class can be derived more than once by different types. Can you explain to me why we are looking for (I think) just any implementation and then filling in the vtable with it? Are abstract classes even supposed to have populated vtables? This is something I honestly don't know anything about, so I may have misunderstood this completely.

`getVTable` - I think this kind of stuff should probably be a pre-calculated property of `TypeInfo` ("VTable"), and it seems like it would make sense to rename `MetadataUsage` to `MetadataUsagePair` and make a new `MetadataUsage` class with this gunk decoding logic as it's repeated code, I would like to see this kind of low-level detail hidden from output modules where possible. The overall behaviour of outputs (and derivative code like GUIs or other applications that include Il2CppInspector as a library) should be that - as much as possible - it should only need to access the `Il2CppModel`, and only go down to the `Il2CppInspector` level if absolutely necessary. Any behaviour that might be desirable to other future output modules should go into `Il2CppModel` or `Il2CppInspector` as appropriate, and I would assume future developers may well be interested in accessing the vtables.

Other than nitpicking, the rest seems basically fine. Whether it actually works correctly and is bug free, I will admit I don't have sufficient knowledge of IL2CPP to answer. But hey, I made the type model precisely so that people could leverage it to make stuff like this!

Regarding a couple of points in your email:

- to prevent duplication, I assume you just have a running list of which specializations have already been output and check against this each time you write a new type? If I've done my job properly, a `TypeInfo` will never be generated twice by `Il2CppModel` for the same type, so you should be able to test this with a simple reference equality. If that's not the case, that's something I should look at and fix.

- IDA structs for every version. I was thinking about the idea of creating Il2CppClass and any other needed structs as actual `unsafe` classes in C#, annotating them with attributes as per the file reading code, and then using reflection to output them correctly. I quasi-concluded that this was unnecessarily over-complicated but hadn't really decided for sure. It's elegant but strange. A text templating approach as you suggested may be better, and I'm open to any reasonable solution. If you do create swathes of text as at present though, please add them as resources as with the Visual Studio solution templates, to keep the source code tidier.

**Comment by nneonneo** [comment]
I’m happy to fix code style nitpicks because I’m going to be making more changes as I go anyway. 

Regarding `populateConcreteImplementations`: this is a silly hack to account for the fact that a vtable for an abstract class has null for any unimplemented abstract methods. Because this makes it impossible to know what method should be in that slot, what I did was to go hunting for *any* concrete implementation of the abstract class, which must have an implementation for all abstract methods. Then we can grab the name and signature of the method from the concrete implementation (which, AFAIK, must be the same as the abstract method). Basically: it’s a way to get name and signature info for abstract methods in the vtable. 

Regarding duplication: it’s not an issue yet, but at some point I’m going to need fleshed out generic classes in order to generate type structs. This includes the base type and other type-parameter-dependent constructed types. The difficulty is that a single concrete (closed) generic type might be reached in more than one way, and deduplicating them is hard. Simple example:

```
class A<T> {}
class B<T1, T2> : A<T2> {}
class C { B<string, int> b; A<int> a; }
```

In order to generate structs for `B<string, int>` we need to generate a concretized version of the base class `A<int>`, but the way that looks in the TypeRef/TypeDef structures is a partial specialization `A<T2>` specialized with args `T1=string, T2=int` which looks distinct from the direct `A<int>` specialization (`A<T>` with T=int), yet yields the same concrete result.

I *think* I can solve this by caching fully closed generic types separately and keying on `(Definition, TypeRefIndex[])`, but it’s kind of an ugly solution.

**Comment by djkaty** [comment]
> Regarding `populateConcreteImplementations`: ... Basically: it’s a way to get name and signature info for abstract methods in the vtable.

Ahh ok, for some reason I thought you were using method pointers in the abstract class which confused me. I'm sure you're aware but just a note that derived classes of an abstract class can themselves be abstract and only implement some of the methods so you may have to iterate down the tree (if you haven't already done this, I didn't check). Implementations will have `override` naturally, there may be co-variant return types at play, and of course be wary of method hiding with `new`. I can't think of any other variations off the top of my head. You must also determine which methods in an implementing class are inherited from the abstract parent class(es) and which aren't. Again I apologize for not looking at the code thoroughly if you have this all in hand already. Thanks for the explanation!

> Regarding duplication: it’s not an issue yet, but at some point I’m going to need fleshed out generic classes in order to generate type structs. This includes the base type and other type-parameter-dependent constructed types.

Are you referring to the TODO in TypeInfo where I wrote that it was a bare class at the moment?

> The difficulty is that a single concrete (closed) generic type might be reached in more than one way, and deduplicating them is hard. Simple example:
> 
> ```
> class A<T> {}
> class B<T1, T2> : A<T2> {}
> class C { B<string, int> b; A<int> a; }
> ```
> 
> In order to generate structs for `B<string, int>` we need to generate a concretized version of the base class `A<int>`, 

Was this:
>but the way that looks in the TypeRef/TypeDef structures is a partial specialization `A<T2>` specialized with args `T1=string, T2=int`

a typo? Is it possible in the metadata for a generic type with a single type parameter to have two type arguments assigned to it and one discarded?

My understanding of type placement in the model is that it should be like this:

- Everything that's a type definition (class/struct) that you see in the C# stub output is in `TypesByDefinitionIndex`. In your example above, this would be `A<T>`, `B<T1, T2>` and `C`.

- Everything that is referred to in said definitions, including base class derivation after the colon, fields and methods are in `TypesByReferenceIndex`. In your example, this would be `A<T2>`, `B<string, int>` and `A<int>`. These will definitely all be present.

- Where things get more sketchy is when we're talking about concrete instances that aren't defined in the source code directly but are instantiated eg. by `new`. So for example if the application with the types above had a `var foo = new A<float>();`, there will be a set of functions for this specialization. These can be found in `TypesByMethodSpecClassIndex`. The deal is the same with concrete generic methods, these can be found in  GenericMethods` (I should probably rename one of these to make them consistent).

You can drill down further and I'd recommend you check out `Il2CppModel.GetGenericMethod` (see `TestGenerics.cs` for usage). Getting the correct generic method is a bit of a pain so I made this fudge function, this is a weakness in the current model that I need to address at some point.

Just to make your life worse I thought I should throw this in in case you come across it and it causes any confusion:

https://blogs.unity3d.com/2015/06/16/il2cpp-internals-generic-sharing-implementation/

You can change whether generic sharing is enabled or disabled during compilation with a flag to `il2cpp.exe`, so definitely experiment with this when you have the opportunity.

Great work so far, I'm thoroughly excited to see this come to fruition!

**Comment by nneonneo** [comment]
> Ahh ok, for some reason I thought you were using method pointers in the abstract class which confused me. I'm sure you're aware but just a note that derived classes of an abstract class can themselves be abstract and only implement some of the methods so you may have to iterate down the tree (if you haven't already done this, I didn't check). Implementations will have override naturally, there may be co-variant return types at play, and of course be wary of method hiding with new. I can't think of any other variations off the top of my head. You must also determine which methods in an implementing class are inherited from the abstract parent class(es) and which aren't. Again I apologize for not looking at the code thoroughly if you have this all in hand already. Thanks for the explanation!

I am walking all classes, selecting *concrete* classes and mapping any of their abstract bases to them. Therefore, I should be able to pick up concrete implementations through multiple layers of abstract bases. I understand the complexity of `override` vs. `new`, and I *believe* that none of those things affect the actual vtable slots - that is, any given vtable slot from an abstract base can *only* be filled with a matching override from a concrete implementation (and, assuming any concrete implementation exists, *must* be filled with such a matching override). So, luckily, all the hard work is already done by Il2Cpp when it figures out how to fill the vtable slots.

> Are you referring to the TODO in TypeInfo where I wrote that it was a bare class at the moment?

Yep. Currently I don't fill in any structures for closed generic types, nor do I do function types for closed generic methods. This is a major limitation right now. 

> a typo? Is it possible in the metadata for a generic type with a single type parameter to have two type arguments assigned to it and one discarded?

Not a typo. Check the `BaseType` (the TypeRef) for `B<T1, T2>` and you'll see it's actually shown as `A<T2>`, not `A<T>`. This is because the TypeRef specifies a partial specialization (GenericImpl/GenericInst) of `A<T>` where `T=T2` (the `T2` of B). Later, when B is fully specialized using e.g. `T1=string, T2=int`, it will specialize its base class via the same args (the same generic context). This works because `A<T2>` uses B's `T2` as the type parameter, which has the correct generic parameter index of 1.

Note that here, `A<T>` would be the original type definition, whereas `A<T2>` would be a type reference.

> Where things get more sketchy is when we're talking about concrete instances that aren't defined in the source code directly but are instantiated eg. by new. So for example if the application with the types above had a var foo = new A<float>();, there will be a set of functions in the compiled code for this specialization. These concrete specializations can be found in TypesByMethodSpecClassIndex. The deal is the same with concrete generic methods, these can be found in GenericMethods (I should probably rename one of these to make them consistent). To re-iterate for clarity: only concrete specializations that are NOT defined as fields, parameters or types etc. directly are in this list, ie. only those that are actually instantiated by source code statements. Definitions and declarations are in the other two lists.

The way libil2cpp appears to handle this is as follows: if a type needs to be specialized at runtime, it pulls out the base `TypeRef` for the specialization (which might itself be partially specialized already) and specializes it using the `argv` from the corresponding `GenericInst` (indexed using `classIndexIndex` from the MethodSpec). Then it recursively specializes everything in the type using the same `argv`.

I've actually renamed both `TypesByMethodSpecClassIndex` and `GenericMethods` in my most recent set of commits; they're now `TypesByGenericInstIndex` and `MethodsByReferenceIndex` respectively. `TypesByGenericInstIndex` is a dictionary mapping `<TypeInfo, GenericInst>` pairs to partially or fully specialized `TypeInfo`s, whereas `MethodsByReferenceIndex` is indexed by `MethodRef` indices (which are also `MethodSpec` indices).

> You can drill down further and I'd recommend you check out Il2CppModel.GetGenericMethod (see TestGenerics.cs for usage). Getting the correct generic method is a bit of a pain so I made this fudge function, this is a weakness in the current model that I need to address at some point.

That is indeed a bit of a fudge. AFAIU the correct approach is to make it possible to instantiate a `MethodBase` (or `MethodInfo`, `ConstructorInfo`) using type + method arguments, or more succinctly a partially/completely specialized parent type (which integrates type arguments) plus method arguments. I tweaked the constructors in my latest patch to do this, which means that `MethodsByReferenceIndex` now contains non-generic methods of generic types.

The best approach would be to actually flesh out generic type instances instead of leaving them bare.

> Just to make your life worse I thought I should throw this in in case you come across it and it causes any confusion:

> https://blogs.unity3d.com/2015/06/16/il2cpp-internals-generic-sharing-implementation/

I read about the Mono implementation of generic sharing over the weekend, saw the relevant structures in Il2Cpp (rgctx*), but didn't realize they'd blogged about it. Thanks for the pointer, will definitely read.

Generics sharing is another minefield I don't plan on walking into right now. I *could* populate the rgctx structures to resolve shared-generic method calls, since I do understand them now, but it's not a priority (and will probably not be done by the time we merge the first version of this, since it affects a small minority of classes right now).

**Comment by nneonneo** [comment]
Oh, before I forget: is there any nice way to figure out if the compiler was MSVC or GCC/Clang, besides simply guessing that all `PE` files must be MSVC? This matters because the way they lay out inherited classes is *very* different, so I need to generate IDA structures that match the compiler.

If you're wondering why I don't just use the field offsets - well, they (a) don't work for generics, (b) are occasionally zero for no good reason in my test files even in concrete classes, and (c) are totally incompatible with IDA's C parser, so I'd have to drop down to IDA's struct API and I definitely do not want to do that.

**Comment by djkaty** [comment]
Read your remarks, thanks for the detailed reply, will reply properly later.

I'm not aware of any specific flag or property for the compiler but I'm sure there must be some fairly obvious signatures we can come up with for that. Would need to do some research. The code that finds CodeRegistration and MetadataRegistration essentially performs a very basic de-compilation of specific code, this code differs quite widely between compilers so there are quite a few implementations (Architectures folder; the older ones are a mess, the newer ones are much more robust and flexible). I don't actually know which implementation matches which compiler but if we can find out, that would be an easy way in to note the compiler used.

Don't worry about the generic sharing :) Unity has an entire blog series called 'IL2CPP Internals' which you can find the rest of from the link. It's from 2015 so some of it is out of date now.

I wanted to point you to another set of articles while it was on my mind: Jackson Dunstan has written quite the trove of excellent articles on Il2CPP: https://jacksondunstan.com/articles/tag/il2cpp - note there are more than this because he has forgotten to tag a few of them, so you might want to dig in to other tags or trawl the list of posts to find the rest. I haven't actually read them all yet but I learned a ton from his excellent work.

**Comment by djkaty** [comment]
I haven't had the opportunity to look at your newer commits yet but a couple of notes on yesterday's chat:

- I'm totally fine with the type list/dictionary properties in the model being renamed, all good

- I will make fleshing out the generic instantiation classes (the recursive type argument substitution that needs to be implemented) the priority as the next thing I work on. I can't promise when this will be done, sometimes I have lots of time to work on this project, sometimes none, but I will work on it next in any case.

- In terms of that fudge method, in the real .NET version of `TypeInfo` they have a `GetGenericTypeDefinition` property, and in `MethodInfo` they have `GetGenericMethodDefinition` which return the completely un-specialized versions of a type or method from a specialized version. Additionally in `MethodInfo` they have `MakeGenericMethod` which allows you to substitute the type parameters of generic method to make a concrete method.

My plan is to use a similar layout to this, with `GetGenericTypeDefinition` and `GetGenericMethodDefinition`, plus properties in `TypeInfo and `MethodInfo` which will return lists of all the specialized versions of a given type or method in an IL2CPP binary (ie. the opposite end of the relationship). That way it will more or less mirror the .NET Reflection API in line with the rest of the current API surface.

**Comment by nneonneo** [comment]
I’ve added GetGenericTypeDefinition already; hopefully the implementation of that is correct? I also split up the getVTable implementation and stuck the low-level bits in the Il2CppInspector class. 

My next steps will be to add function types to the VTable, which will further help decompilation. 

Robert

**Comment by djkaty** [comment]
I've reviewed all of the code now, at least skimmed it. I'll give it another once over when the final PR is ready.

Most of it looks great, you've done a lot of great work here!

Some questions/comments:

* Due to the size of this PR I will do a rebase onto master rather than a single merge commit, so try to keep your commits restricted in scope. The main thing I noticed was the lumping in of GetGenericTypeDefinition() with other code in the same commit. The rest was fine I think, just a note for future reference

* Can you ELI5 why `GenericMethods` (as it was before) has to contain non-generic methods of generic types?

* Don't create `TypeInfo` objects outside of `Il2CppModel`. The idea is to have all type creation completed by the time the constructor of `Il2CppModel` returns so that the entire type model is available to all consumers

* Implementation of `GetGenericTypeDefinition()` seemed fine as long as you have populated (or null'd) it for every possible constructor path of `TypeInfo`?

* Why was the handling of `IL2CPP_TYPE_GENERICINST`? My assumption is to avoid duplicates but perhaps you can confirm?

* `StringLiteral_` is the prefix I have seen of IL2CPP compilations that have symbol tables included. I don't know if this is universal but I'd prefer it to be kept like that.

* Do you think you could add a `VTable` property to `TypeInfo` that is just a surrogate to the methods you've already made? `Il2CppInspector` should generally be de-coupled from `Il2CppModel` in that the former should (ideally) not have to know about the existence of the latter, although I realize that is very difficult to achieve in practice and I have not yet achieved that 100% either. Generally that means `Il2CppInspector` doesn't get involved in methods receiving or returning types from the type model like `MethodBase`.

* If you'd like to go ahead and fill in the type parameter substitution code in the `TypeInfo` constructor we talked about earlier, feel free to do that. It will likely be quite difficult as you can't just create more `TypeInfo`s, you need to search the existing ones for matches to avoid creating duplicates, but at this point it seems like you are more on top of the intricacies of how this works than I am. If you don't want to do it that's totally fine of course!

* I left my main question/confusion until last. In an earlier commit before you made further changes, you noted that the old `TypesByMethodSpecClassIndex` list was in accurate. In the code you added to the IDA script output, you then re-created or accessed the same types using the same code that was populating the list in the first place, just without storing them in the list. I did not understand the difference? The grouping by `GenericInst` confused the heck out of me despite reading your extended commit messages and good comments. Why does the list property not work and why can duplicates not be resolved by (admittedly inefficiently) iterating `TypesByMethodSpecClassIndex.Values` for matching `TypeInfo`s? If there is a match, the new entry can be added as a reference to said `TypeInfo`.

On this last question it's very likely that I'm misunderstanding the relationships between the metadata tables I suspect, so a clarification would be great.

Sorry for so many remarks, this is a major change to the code base that warrants a new version release in and of itself.

**Comment by djkaty** [comment]
Follow-up: all of the test binaries in the test suite are now failing with `IndexOutOfRangeException` in `ParameterInfo.ParameterType`. Seems like a problem with ` TypesByReferenceIndex`.

`TestGenerics.cs` is failing with a signature mismatch error on `GenericMethodDefinitioninGenericClass2<U>(T a, U b)`. `U` is expected to be `string` and `T` is an unspecialized generic type parameter of `GenericClassWithMethods<T>` but the code appears to be using the `GenericClassWithMethods<int>` specialization.

Please check that all the unit tests pass :)

**Comment by nneonneo** [comment]
> I've reviewed all of the code now, at least skimmed it. I'll give it another once over when the final PR is ready.
> 
> Most of it looks great, you've done a lot of great work here!
> 
> Some questions/comments:
> 
>     * Due to the size of this PR I will do a rebase onto master rather than a single merge commit, so try to keep your commits restricted in scope. The main thing I noticed was the lumping in of GetGenericTypeDefinition() with other code in the same commit. The rest was fine I think, just a note for future reference

OK. I can also split that commit with an interactive rebase. Thanks for catching that - I'm usually a little better with `git add -p` :P

>     * Can you ELI5 why `GenericMethods` (as it was before) has to contain non-generic methods of generic types?

When `methodIndexIndex=-1` in `MethodSpec`, it means that the method itself is not generic (it has no method generic arguments), but the type it's in does. This means that the method may have generic arguments from the type, which means that it can be specialized once `classIndexIndex` is used to specialize the containing type.

Also, the original code had this comment in it:

> List of all methods from MethodSpecs (closed generic methods that can be called; does not need to be in a generic class)

but previously the GenericMethods was omitting all methods which had `methodIndexIndex = -1`. These methods are very common in the binary (in fact, probably *most* concrete generic methods are of this type), so adding them resulted in substantially improved symbol coverage for functions in the binary.

>     * Don't create `TypeInfo` objects outside of `Il2CppModel`. The idea is to have all type creation completed by the time the constructor of `Il2CppModel` returns so that the entire type model is available to all consumers

`GetGenericInstType` currently instantiates `TypeInfo`, but in theory every call to this method will be completed by the time we finish initializing `Il2CppModel`. A previous commit created TypeInfo structs in the IDA script generator, but that was a temporary hack until I changed the generic type instance code; the IDA script generator should no longer be creating TypeInfos.

>     * Implementation of `GetGenericTypeDefinition()` seemed fine as long as you have populated (or null'd) it for every possible constructor path of `TypeInfo`?

I think so - feel free to double check. I don't know what it returns in the case of array/ref generics though, so maybe we need a generic type definition there.

>     * Why was the handling of `IL2CPP_TYPE_GENERICINST` moved? My assumption is to avoid duplicates but perhaps you can confirm?

Yes, to avoid duplicates. We can't dedupe in a constructor, unfortunately, and a generic instance can be instantiated in two places (via TypeRef and via MethodSpec, and there are a handful of MethodSpec types which don't appear in a TypeRef). In the future this will also include specialization of the base type of a generic type.

>     * `StringLiteral_` is the prefix I have seen of IL2CPP compilations that have symbol tables included. I don't know if this is universal but I'd prefer it to be kept like that.

OK, I can change that. However, I'd like to keep the actual content of the string in the name; IDA decompiles strings normally this way (e.g. the string "Hello world" is given the auto-name `aHelloWorld`), and it makes the decompilation *significantly* easier to read. 

>     * Do you think you could add a `VTable` property to `TypeInfo` that is just a surrogate to the methods you've already made? `Il2CppInspector` should generally be de-coupled from `Il2CppModel` in that the former should (ideally) not have to know about the existence of the latter, although I realize that is very difficult to achieve in practice and I have not yet achieved that 100% either. Generally that means `Il2CppInspector` doesn't get involved in methods receiving or returning types from the type model like `MethodBase`.

Hopefully I dealt with that here: https://github.com/djkaty/Il2CppInspector/pull/24/files#diff-912f4a26192125f23a3f3498e78e95d2R232. I split the VTable code into two bits: the bit in `Il2CppInspector` returns an array of `MetadataUsage`, and the bit in `Il2CppModel` instantiates that into an array of `MethodBase`. Happy to take suggestions on an improved design.

>     * If you'd like to go ahead and fill in the type parameter substitution code in the `TypeInfo` constructor we talked about earlier, feel free to do that. It will likely be quite difficult as you can't just create more `TypeInfo`s, you need to search the existing ones for matches to avoid creating duplicates, but at this point it seems like you are more on top of the intricacies of how this works than I am. If you don't want to do it that's totally fine of course!

It's unfortunately not that trivial; the base class is the hardest thing to handle due to recursive types (e.g. `class A : B<A>` requires A in the base class, which means you can't specialize in the constructor because the TypeInfo won't be in any tables yet). I haven't prototyped it yet to see how hard it would be, though, because I've been focusing on other parts of the code.

>     * I left my main question/confusion until last. In an earlier commit before you made further changes, you noted that the old `TypesByMethodSpecClassIndex` list was in accurate. In the code you added to the IDA script output, you then re-created or accessed the same types using the same code that was populating the list in the first place, just without storing them in the list. I did not understand the difference? The grouping by `GenericInst` confused the heck out of me despite reading your extended commit messages and good comments. Why does the list property not work and why can duplicates not be resolved by (admittedly inefficiently) iterating `TypesByMethodSpecClassIndex.Values` for matching `TypeInfo`s? If there is a match, the new entry can be added as a reference to said `TypeInfo`.

This should be totally fixed in the new commits. The construction of TypeInfo was a temporary hack in the IDA code.

Previously, `TypesByMethodSpecClassIndex` was caching solely by `classIndexIndex`. This doesn't work because `classIndexIndex` is just an index to a `GenericInst`, which is a list of type arguments *without reference to the generic type being instantiated*. To give an example, suppose you have `class A<T>`, `class B<T>`, and two instantiations `A<int>` and `B<int>`. Despite being unrelated, Il2Cpp is allowed to use a single `GenericInst` containing `[int]` as the type arguments to both. This means that, when looking up the type `A<int>` by its `classIndexIndex` in `TypesByMethodSpecClassIndex`, you might end up with `B<int>` instead (and this did actually happen when generating symbol names).

`TypesByGenericInst` indexes by both the parent generic type and by the `GenericInst`, which resolves this. (It's not perfect - it can contain duplicates if the parent generics are different partial specializations of the same type, e.g. `A<T>` and `A<T2>`, and the GenericInsts contain the same concrete arguments - so we do need a caching mechanism for fully concrete types...)

> On this last question it's very likely that I'm misunderstanding the relationships between the metadata tables I suspect, so a clarification would be great.
> 
> Sorry for so many remarks, this is a major change to the code base that warrants a new version release in and of itself.

No, this is great. I really appreciate you taking the time to look through everything and give your feedback.


> Follow-up: all of the test binaries in the test suite are now failing with `IndexOutOfRangeException` in `ParameterInfo.ParameterType`. Seems like a problem with ` TypesByReferenceIndex`.
> 
> `TestGenerics.cs` is failing with a signature mismatch error on `GenericMethodDefinitioninGenericClass2<U>(T a, U b)`. `U` is expected to be `string` and `T` is an unspecialized generic type parameter of `GenericClassWithMethods<T>` but the code appears to be using the `GenericClassWithMethods<int>` specialization.
> 
> Please check that all the unit tests pass :)

Whoops, I will do that.

OK, so I don't forget, here are my concrete TODOs:

[ ] Clean up my commits a little bit so they can be cleanly rebased with good history (mostly splitting some commits, maybe rebasing over the silly hack code in the IDA script generator that I immediately overwrote)
[ ] Double check `GetGenericTypeDefinition` and try to match the real one as much as possible (wrt. ref/ptr/array derived types)
[ ] Change `strLit` back to `StringLiteral`
[ ] Fix unit tests

And so you're aware of my personal TODO list for this PR:

[ ] add MethodInfo argument to all function calls (will require MethodInfo struct)
[ ] refactor to move Il2Cpp base struct definitions to their own file
[ ] add support for older Il2Cpp versions
[ ] correctly output field structures for GNU/Clang compilers (requires flattening the field list)
[ ] add types for VTables so that virtual method calls get typechecked in IDA (*especially* the return types)

Stretch goals:

[ ] Parameter substitution/concretization of generic TypeInfo
[ ] rgctx structures (I understand these, just need to hook them up to `MethodInfo`/`Il2CppClass` subclasses) - they're occasionally used in the code
[ ] type the data segment of the binary using the Il2Cpp structs (this is low priority, even though it's a todo in the code, because it's not that important for reversing; it just helps identify segments of the binary we can ignore)

**Comment by djkaty** [comment]
Just a quick note to say I haven't forgotten, I got sick and have been in isolation for 8 days and have had to focus on other things. I will get back to this as soon as I'm able to. Apologies for the delay.

**Comment by djkaty** [comment]
Is this PR to be closed without merging?

**Comment by nneonneo** [comment]
Yes, it can be closed. I will be posting the final PR later this week. 

Robert
